Add RefreshMode and Times to Java Gateway

mschmiedel opened the pull request:
Idea: avoid sending too much data over the network. Setting a polling interval will reduce traffic, normally 30 seconds resolution should be enough. 

Ideas: 
* set a value > 0 as default, however this changes current behavior, maybe expectations of certain users.
* limit the value to a maximum. Is there any risk of the car becoming unresponsive while HV battery is active? Maybe setting too high values (like 10 minutes) kills the purpose of the gateway?

Tonno87 commented:
As described here [(issue 46)](https://github.com/ReverseEngineeringDE/SAIC-API-Documentation/issues/46) the polling time should be  implemented by MQTT (refresh/period/active or refresh/period/InActive). So it's not necessary to implement a commandline option.

mschmiedel commented:
OK, means that functionality is right, but should base on that topics instead of parameter? Could change that in the afternoon.

Tonno87 commented:
Both Gateways should work the same way, so take a look at [saic-python-mqtt-gateway](https://github.com/SAIC-iSmart-API/saic-python-mqtt-gateway).
The idea is that the gateway does nothing without a motor start message or forced updates over mqtt. It only looks for these start conditions, but without it there is only one poll per day (**inActive** value). But the **active** value defines how often you get an update while driving (30s is default).
Additionally you can change the mode of polling with the topic **saic/xxx/vehicles/xxx/refresh/mode/set**:
- off
- periodic (like auto mode)
- force

Under these conditions the 12V battery drain should be minimized enough :)

On the python gateway there is a third way to poll: OpenWB Loadpoint Map. But i think this is not planned for java.

So there is still a lot to do...

mschmiedel commented:
Hi,

I switched the logic to the behavior I understood from the python gateway. 

However, what I don't like, it seems as if restarting the gateway will also reset the values to the default (30s/1day). Is that correct and expected? I would rather like to check if there is a value in the topic and use it, or else set the default...

mschmiedel commented:
I have been wanting to test the feature directly on the vehicle, but API seems broken right now, not sure if it only affects my car or the API in general. I would be OK to wait until it is completely testable...

tisoft commented:
Thanks for the PR, I'll try to test it next weekend. :)

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -460,9 +474,19 @@ public void notifyMessage(SaicMessage message) throws MqttException {
+      case OFF:
+        return false;
+      case FORCE:
+        setRefreshMode(PERIODIC);
```
FORCE should leave the previous refresh mode as is, so that you can set it to OFF for no automatic refreshes, but still send a FORCE from time to time for a manual update

mschmiedel replied:
OK, that is a good point, will check it out, I assumed periodic, but OFF is also a possibility...

mschmiedel commented:
Fixed the FORCE issue, and added some logging in case SAIC API fails to make the behaviour of the gateway more transparent. Also each gateway refresh now causes a log entry (with 30s refresh delay it is manageable). If you consider this to be rather a debug message (LOGGER.info("Refreshing vehicle status...");), then we could change it. In this case I might also suggest switching all the JSON printing stuff to trace so users can use the debug flag to get usable information...

mschmiedel commented:
Added one more commit to better group code by feature...

Added a workaround to keep chargeStatus from being called if vehicleStatus was not succesful.

mschmiedel commented:
Did some more detailed testing and modified some other things. One question I would have is that right now, if we switch from active to inactive mode, it will do it more or less immediately. With the 1day inactive timer, chances are big that charging status will never be caught, as there is no charging start message afaik. Is there a grace period that should be implemented to overlap the active mode for a few minutes after engine stopped?

mschmiedel commented:
> Before this change, I had a grace time from 10 Minutes after engine off (or charging end) was detected. I think we should keep something like that. Maybe also make it configurable?
> 
> Thanks for all your work by the way. I will go over the PRs soon, time is a bit sparse at the moment.

Will check the Python code and compare it, if we set a configurable value, it will need to be set as a topic. Shall I propose one?

mschmiedel commented:
@tisoft , please review the topic name for the 10m tolerance after shutdown as this is a proposal from me, didn´t find it documented anywhere...

tisoft commented:
Lets's discuss the name here: https://github.com/orgs/SAIC-iSmart-API/discussions/6#discussioncomment-5923443

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleHandler.java:
```diff
@@ -375,11 +391,43 @@ public void handleMQTTCommand(String topic, MqttMessage message) throws MqttExce
+              throw new MqttGatewayException("Unsupported payload " + message);
+          }
+          break;
+        case "refresh/mode":
```
Extract constant

mschmiedel replied:
Will do in the following days. In this case, we should create a constants Class that holds all topics to make it more transparent...

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleHandler.java:
```diff
@@ -375,11 +391,43 @@ public void handleMQTTCommand(String topic, MqttMessage message) throws MqttExce
+                    throw new MqttGatewayException("Unsupported payload " + message);
+                  });
+          break;
+        case "refresh/period/active":
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleHandler.java:
```diff
@@ -375,11 +391,43 @@ public void handleMQTTCommand(String topic, MqttMessage message) throws MqttExce
+            throw new MqttGatewayException("Error setting value for payload: " + message);
+          }
+          break;
+        case "refresh/period/inActive":
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleHandler.java:
```diff
@@ -375,11 +391,43 @@ public void handleMQTTCommand(String topic, MqttMessage message) throws MqttExce
+            throw new MqttGatewayException("Error setting value for payload: " + message);
+          }
+          break;
+        case "refresh/period/afterShutdown":
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -455,17 +472,48 @@ public void notifyMessage(SaicMessage message) throws MqttException {
+    if (lastCarActivity.isAfter(lastSuccessfulRefresh)) {
+      return true;
+    }
+    if (apiUpdateError > 10) {
```
If we get over 10 api errors, can they ever be reset?

mschmiedel replied:
Yes, as soon as the car sends a message to the api or if before the 10 apiErrors it refreshes correctly. However, if the gateway goes into a non recoverable api error state, it will be doomed anyway... ;) Or do you have a better solution for this? Idea is that a failing api will not bomb the api.

One other idea is that the if somehow waits for like 1h if > apiErrors is > 10, this way a broken api will not be bombed with requests...

mschmiedel replied:
I will remove this part for now. Not sure if it fits into this MR and is easily reimplemented, else we will be on this MR forever. :)

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -496,4 +544,84 @@ public void configure(VinInfo vinInfo) throws MqttException {
+        new MqttMessage(String.valueOf(refreshPeriodInactive).getBytes(StandardCharsets.UTF_8));
+    try {
+      mqttMessage.setRetained(true);
+      this.client.publish(this.mqttVINPrefix + "/refresh/period/inActive", mqttMessage);
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -496,4 +544,84 @@ public void configure(VinInfo vinInfo) throws MqttException {
+    try {
+      LOGGER.info("Setting refresh mode to {}", refreshMode.getStringValue());
+      mqttMessage.setRetained(true);
+      this.client.publish(this.mqttVINPrefix + "/refresh/mode", mqttMessage);
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -496,4 +544,84 @@ public void configure(VinInfo vinInfo) throws MqttException {
+        new MqttMessage(String.valueOf(refreshPeriodActive).getBytes(StandardCharsets.UTF_8));
+    try {
+      mqttMessage.setRetained(true);
+      this.client.publish(this.mqttVINPrefix + "/refresh/period/active", mqttMessage);
```
Extract constant

tisoft commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleState.java:
```diff
@@ -496,4 +544,84 @@ public void configure(VinInfo vinInfo) throws MqttException {
+            String.valueOf(refreshPeriodAfterShutdown).getBytes(StandardCharsets.UTF_8));
+    try {
+      mqttMessage.setRetained(true);
+      this.client.publish(this.mqttVINPrefix + "/refresh/period/afterShutdown", mqttMessage);
```
Extract constant

tisoft reviewed (commented):
Added some review comments. Haven't tested the code yet, will do so this weekend.

mschmiedel commented on saic-java-mqtt-gateway/src/main/java/net/heberling/ismart/mqtt/VehicleHandler.java:
```diff
@@ -66,9 +66,13 @@ void handleVehicle() throws MqttException, IOException {
         OTA_RVMVehicleStatusResp25857 vehicleStatus =
             updateVehicleStatus(uid, token, vinInfo.getVin());
+        if (vehicleStatus == null) {
+          continue; // TODO find better solution for this...
```
It was already done, but improved the handling there with Exception handling...

tisoft commented:
The topic constants sometimes start with a `/` and sometime not, same for their ending. I think we should remove all leading and trailing `/` in the constants and handle that in the place they are used.

Currently the topics for `refresh`, `drivetrain` and `climate`are broken, because they are missing a `/`.

tisoft commented:
@mschmiedel I rebased the commits a bit and squashed the format/lint into the corresponding commits.

Will do a final review tomorrow and then do the merge.

tisoft commented:
@mschmiedel Vielen Dank für den PR!
